### PR TITLE
[integration]: Mem0 as memory

### PR DESCRIPTION
**Pull request by spike-spiegel-21**
# LlamaIndex Memory Integration: Mem0

## Installation

To install the required package, run:

```bash
%pip install llama-index-memory-mem0
```

## Setup with Mem0 Platform

1. Set your Mem0 Platform API key as an environment variable. You can replace `<your-mem0-api-key>` with your actual API key:

> Note: You can obtain your Mem0 Platform API key from the [Mem0 Platform](https://app.mem0.ai/login).

```python
os.environ["MEM0_API_KEY"] = "<your-mem0-api-key>"
```

2. Import the necessary modules and create a Mem0Memory instance:

```python
from llama_index.memory.mem0 import Mem0Memory

context = {"user_id": "user_1"}
memory = Mem0Memory.from_client(
    context=context,
    api_key="<your-mem0-api-key>",
)
```

## Setup with Mem0 OSS

1. Set your Mem0 OSS by providing configuration details:

> Note: To know more about Mem0 OSS, read [Mem0 OSS Quickstart](https://docs.mem0.ai/open-source/quickstart).

```python
config = {
    "vector_store": {
        "provider": "qdrant",
        "config": {
            "collection_name": "test_9",
            "host": "localhost",
            "port": 6333,
            "embedding_model_dims": 1536,  # Change this according to your local model's dimensions
        },
    },
    "llm": {
        "provider": "openai",
        "config": {
            "model": "gpt-4o",
            "temperature": 0.2,
            "max_tokens": 1500,
        },
    },
    "embedder": {
        "provider": "openai",
        "config": {"model": "text-embedding-3-small"},
    },
    "version": "v1.1",
}
```

2. Create a Mem0Memory instance:

```python
memory = Mem0Memory.from_config(
    context=context,
    config=config,
)
```

## Mem0 Context

Mem0 Context is used to identify the user, agent or the conversation in the Mem0. It is required to be passed in the at least one of the fields in the `Mem0Memory` constructor.

```python
context = {
    "user_id": "user_1",
    "agent_id": "agent_1",
    "run_id": "run_1",
}
```

## Basic Usage

Currently, Mem0 Memory is supported in the `SimpleChatEngine`, `FunctionCallingAgent` and `ReActAgent`.

Intilaize the LLM

```python
import os
from llama_index.llms.openai import OpenAI

os.environ["OPENAI_API_KEY"] = "<your-openai-api-key>"
llm = OpenAI(model="gpt-4o")
```

### SimpleChatEngine

```python
from llama_index.core import SimpleChatEngine

agent = SimpleChatEngine.from_defaults(
    llm=llm, memory=memory  # set you memory here
)

# Start the chat
response = agent.chat("Hi, My name is Mayank")
print(response)
```

Initialize the tools

```python
from llama_index.core.tools import FunctionTool


def call_fn(name: str):
    """Call the provided name.
    Args:
        name: str (Name of the person)
    """
    print(f"Calling... {name}")


def email_fn(name: str):
    """Email the provided name.
    Args:
        name: str (Name of the person)
    """
    print(f"Emailing... {name}")


call_tool = FunctionTool.from_defaults(fn=call_fn)
email_tool = FunctionTool.from_defaults(fn=email_fn)
```

### FunctionCallingAgent

```python
from llama_index.core.agent import FunctionCallingAgent

agent = FunctionCallingAgent.from_tools(
    [call_tool, email_tool],
    llm=llm,
    memory=memory,
    verbose=True,
)

# Start the chat
response = agent.chat("Hi, My name is Mayank")
print(response)
```

### ReActAgent

```python
from llama_index.core.agent import ReActAgent

agent = ReActAgent.from_tools(
    [call_tool, email_tool],
    llm=llm,
    memory=memory,
    verbose=True,
)

# Start the chat
response = agent.chat("Hi, My name is Mayank")
print(response)
```

> Note: For more examples refer to: [Notebooks](https://github.com/run-llama/llama_index/tree/main/docs/docs/examples/memory)

## References

- [Mem0 Platform](https://app.mem0.ai/login)
- [Mem0 OSS](https://docs.mem0.ai/open-source/quickstart)
- [Mem0 Github](https://github.com/mem0ai/mem0)

### CODE REVIEW

File: docs/docs/examples/memory/llama_index_mem0_chat_engine.ipynb
Comment (logan-markewich) [inline docs/docs/examples/memory/llama_index_mem0_chat_engine.ipynb:1]: Can we add the installs needed to run this? `%pip install llama-index-memory-mem0 ...` right?

File: llama-index-integrations/memory/llama-index-memory-mem0/README.md
Comment (logan-markewich) [inline llama-index-integrations/memory/llama-index-memory-mem0/README.md:1]: Do you mind adding some more info to this readme?

Like maybe the install command and basic usage?

File: llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/mem0/base.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    
+    @model_validator(mode='after')
+    def check_at_least_one_id(cls, values):
+        if not any([values.user_id, values.agent_id, values.run_id]):
+            raise ValueError("At least one of 'user_id', 'agent_id', or 'run_id' must be provided!")
+        return values
+
+
+class Mem0(BaseMemory):
+    class Config:
```
Comment (logan-markewich) [inline llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/mem0/base.py:29]: Actually, in pydanticv2, the preferred way is `model_config = ConfigDict(...)` 
https://docs.pydantic.dev/latest/concepts/config/

File: llama-index-integrations/memory/llama-index-memory-mem0/tests/__init__.py
Comment (logan-markewich) [inline llama-index-integrations/memory/llama-index-memory-mem0/tests/__init__.py:1]: Any chance we can add one or two tests?

File: llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/mem0/base.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        if isinstance(self.client, Memory):
+            responses = responses['results']
+        response_messages = [self._convert_memory_to_system_message(responses)]
+        # chat engine not sending latest message
+        response_messages.append(self.buffer_user_message)
+        return response_messages
+
+    def put(self, message: ChatMessage) -> None:
+        #TODO: Remove the repetitive memory
+        if message.role == MessageRole.USER:
```
Comment (logan-markewich) [inline llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/mem0/base.py:126]: Only user messages can be inserted?

**Comment by spike-spiegel-21** [comment]
Hi @logan-markewich , noted all the observations. The PR is still under development. Mem0 team is in touch with me, I will mention you once it is done. Thank you for your patience.

File: llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/mem0/base.py (hunk 2 above)
Comment (logan-markewich) [reply llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/mem0/base.py:126]: How do we keep track of responses from the LLM ?

**Comment by logan-markewich** [comment]
@spike-spiegel-21 ah great! Didn't realize you were still working on it, sounds good

**Comment by spike-spiegel-21** [comment]
Hi @logan-markewich , Could you please review this integration?

**Comment by spike-spiegel-21** [comment]
Closing this, will be raising another one

**Pull request closed without merging**